File: Projetos_C/Terceiro Semestre/Estrutura de Dados/Aula 11/Ex3.c
/*Elabore um método para manter duas pilhas dentro de um único vetor linear de tamanho limitado
(STACKSIZE), de modo que nenhuma das pilhas alerte estouro até toda a memória seja usada; e
uma pilha inteira nunca seja deslocada para outro lugar dentro do vetor. Escreva rotinas push1,
pop1, push2 e pop2 para manipular as duas pilhas.*/

#include<stdio.h>
#include<stdlib.h>
#define STACKSIZE 10

struct stack {
	int top1;
    int top2;
	int items[STACKSIZE];
};

int empty1 (struct stack *ps);
int pop1 (struct stack *ps);
void push1 (struct stack *ps, int x);
int pop2 (struct stack *ps);
void push2 (struct stack *ps, int x);

int main(void){
    struct  stack pilha={};
    pilha.top1=-1;
    pilha.top2=STACKSIZE;

    for(int i=0; i<STACKSIZE/2; i++){
        push1(&pilha, i);
        push2(&pilha, i);
    }
    for(int i=0; i<STACKSIZE/2; i++){
        pop1(&pilha);
        pop2(&pilha);
    }

    
    return 0;
}

int empty1 (struct stack *ps){
	if (ps->top1==-1)
		return 1;
	return 0;	
}

int pop1 (struct stack *ps){
	int aux;
	if (empty1(ps)){
		printf("ERRO- PILHA VAZIA");
		exit(1);
	}
	
	aux=ps->items[ps->top1];
	ps->top1--;
	return aux;
}

void push1 (struct stack *ps, int x){
	if (ps->top1==ps->top2){
		printf("ERRO- PILHA CHEIA");
		exit(1);		
	}
	ps->top1++;
	ps->items[ps->top1]=x;
}

int empty2 (struct stack *ps){
	if (ps->top2==STACKSIZE)
		return 1;
	return 0;	
}


int pop2 (struct stack *ps){
	int aux;
	if (empty2(ps)){
		printf("ERRO- PILHA VAZIA");
		exit(1);
	}
	
	aux=ps->items[ps->top2];
	ps->top2++;
	return aux;
}

void push2 (struct stack *ps, int x){
	if (ps->top1==ps->top2){
		printf("ERRO- PILHA CHEIA");
		exit(1);		
	}

	ps->top2--;
	ps->items[ps->top2]=x;
}

File: Projetos_Python/3 Semestre/Primeira Fase/Prova1/3-Pilha e Fila/Lista1/Ex3.py
"""
3. Elabore um método para manter duas pilhas dentro de um único vetor linear de tamanho limitado
(STACKSIZE), de modo que nenhuma das pilhas alerte estouro até toda a memória seja usada; e 
uma pilha inteira nunca seja deslocada para outro lugar dentro do vetor. Escreva rotinas push1, 
pop1, push2 e pop2 para manipular as duas pilhas.
"""
STACKSIZE=100
class Pilha:

  def __init__(self):
    self.topo1 = -1
    self.topo2=STACKSIZE
    self.valores = [None]*STACKSIZE

  def pilha_cheia(self):
    if self.topo1 + 1 == self.topo2:
      return True
    return False

  def pilha_vazia1(self):
    if self.topo1 == -1:
      return True
    return False
  
  def pilha_vazia2(self):
    if self.topo2 == STACKSIZE:
      return True
    return False

  def empilhar1(self, valor):
    if self.pilha_cheia():
      print('A pilha está cheia')
      exit(1)
    self.topo1 += 1
    self.valores[self.topo] = valor

  def desempilhar1(self):
    if self.pilha_vazia1():
      print('A pilha está vazia')
      exit(1)
    aux=self.valores[self.topo1]
    self.topo1 -= 1
    return aux
  
  def empilhar2(self, valor):
        if self.pilha_cheia():
            print('A pilha está cheia')
            exit(1)
        self.topo2 -= 1
        self.valores[self.topo2] = valor
  
  def desempilhar2(self):
    if self.pilha_vazia2():
      print('A pilha está vazia')
      exit(1)
    aux=self.valores[self.topo2]
    self.topo2 += 1
    return aux
  


